File: java/src/main/java/sbuch/presentation/examples/TrafficLightExample.java
package sbuch.presentation.examples;

import java.util.logging.Logger;


public class TrafficLightExample {
    private static final Logger LOGGER = Logger.getLogger(TrafficLightExample.class.getName());

    public static void main(String[] args) {
        logTrafficLightState(TrafficLightColor.RED);
        logTrafficLightState(TrafficLightColor.ORANGE);
        logTrafficLightState(TrafficLightColor.GREEN);
    }

    private static void logTrafficLightState(TrafficLightColor color) {
        switch (color) {
            case RED -> LOGGER.info("Red");
            case ORANGE -> LOGGER.info("Orange");
            case GREEN -> LOGGER.info("Green");
        }
    }
}

// TrafficLight = (Boolean, Boolean, Boolean)
record TrafficLight(Boolean redOn,
                    Boolean orangeOn,
                    Boolean greenOn) {
}

// TrafficLightColor = String
// Use String

// TrafficLightColor = "red" | "orange" | "green"
enum TrafficLightColor {RED, ORANGE, GREEN}

File: typescript/src/trafficLightExample.ts
// TrafficLight = (Boolean, Boolean, Boolean)
// type TrafficLight = [boolean, boolean, boolean];

// TrafficLightColor = String
// type TrafficLightColor = string

// TrafficLightColor = "red" | "orange" | "green"
type TrafficLightColor = "red" | "orange" | "green"

function logTrafficLightState(color: TrafficLightColor): void {
  console.info(`Traffic Light Color: ${color}!`);
}

// This is OK
logTrafficLightState("red");
logTrafficLightState("orange");
logTrafficLightState("green");

// This is also fine
const red = "red";
logTrafficLightState(red);

// This is not OK and tsc won't compile
// logTrafficLightState("violet");
